dss-collection-nft/contracts/DSSCollection.cdc

```diff
@@ -0,0 +1,626 @@
+            pre {
+                DSSCollection.CollectionGroupData(
+                    id: collectionGroupID
+                ).open == true: "collection group not open"
+            }
+
+            self.id = DSSCollection.nextSlotID
+            self.collectionGroupID = collectionGroupID
+            self.logicalOperator = logicalOperator
+            self.typeName = typeName
+            self.slotType = slotType
+            self.items = {}
+
+            DSSCollection.nextSlotID = self.id + 1 as UInt64
+
+            emit SlotCreated(
+                id: self.id,
+                collectionGroupID: self.collectionGroupID,
+                logicalOperator: self.logicalOperator,
+                typeName: self.typeName,
+                slotType: self.slotType
+            )
+        }
+    }
+
+    // A public struct to access CollectionGroup data
+    //
+    pub struct CollectionGroupData {
+        pub let id: UInt64
+        pub let name: String
+        pub let typeName: String
+        pub let open: Bool
+        pub let startTime: UFix64?
+        pub let endTime: UFix64?
+        pub let timeBound: Bool
+
+        init (id: UInt64) {
+            if let collectionGroup = &DSSCollection.collectionGroupByID[id] as &DSSCollection.CollectionGroup? {
+                self.id = collectionGroup.id
+                self.name = collectionGroup.name
+                self.typeName = collectionGroup.typeName
+                self.open = collectionGroup.open
+                self.startTime = collectionGroup.startTime
+                self.endTime = collectionGroup.endTime
+                self.timeBound = collectionGroup.timeBound
+            } else {
+                panic("collectionGroup does not exist")
+            }
+        }
+    }
+
+    // A top-level CollectionGroup with a unique ID and name
+    //
+    pub resource CollectionGroup {
+        pub let id: UInt64
+        pub let name: String
+        pub let typeName: String
+        pub var open: Bool
+        pub let startTime: UFix64?
+        pub let endTime: UFix64?
+        pub let timeBound: Bool
+        pub var numMinted: UInt64
+
+        // Close this collection group
+        //
+        access(contract) fun close() {
+            pre {
+                self.open == true: "not open"
+            }
+
+            self.open = false
+
+            emit CollectionGroupClosed(id: self.id)
+        }
+
+        // Mint a DSSCollection NFT in this group
+        //
+        pub fun mint(completedBy: String, level: UInt64): @DSSCollection.NFT {
+            pre {
+                self.open != true: "cannot mint an open collection group"
+                DSSCollection.validateTimeRange(
+                    timeBound: self.timeBound,
+                    startTime: self.startTime,
+                    endTime: self.endTime
+                ) == true : "cannot mint a collection group outside of time bounds"
+                level <= 10: "token level must be less than 10"
```

Comment:
Super nit:  If you're limiting it to 10, seems like UInt8 is more appropriate.  When I see UInt64, it tells me to expect some large numbers.

Also :(, I was looking forward to a level 2^64 :P.

Reply:
Yes. Have changed to UInt8. Thank you.